resolwe/flow/views/entity.py

```diff
@@ -0,0 +1,139 @@
+        """Destroy a model instance.
+
+        If ``delete_content`` flag is set in query parameters, also all
+        Data objects contained in entity will be deleted.
+        """
+        obj = self.get_object()
+        user = request.user
+
+        if strtobool(request.query_params.get('delete_content', 'false')):
+            for data in obj.data.all():
+                if user.has_perm('edit_data', data):
+                    data.delete()
+
+        # NOTE: Collection's ``destroy`` method should be skiped, so we
+        # intentionaly call it's parent.
+        return super(CollectionViewSet, self).destroy(  # pylint: disable=no-member,bad-super-call
+            request, *args, **kwargs
+        )
+
+    @detail_route(methods=[u'post'])
+    def add_to_collection(self, request, pk=None):
+        """Add Entity to a collection."""
+        entity = self.get_object()
+
+        if 'ids' not in request.data:
+            return Response({"error": "`ids` parameter is required"}, status=status.HTTP_400_BAD_REQUEST)
+
+        for collection_id in request.data['ids']:
+            self._check_collection_permissions(collection_id, request.user)
+
+        for collection_id in request.data['ids']:
+            entity.collections.add(collection_id)
+
+            collection = Collection.objects.get(pk=collection_id)
+            for data in entity.data.all():
+                collection.data.add(data)
+
+        return Response()
+
+    @detail_route(methods=[u'post'])
+    def remove_from_collection(self, request, pk=None):
+        """Remove Entity from a collection."""
+        entity = self.get_object()
+
+        if 'ids' not in request.data:
+            return Response({"error": "`ids` parameter is required"}, status=status.HTTP_400_BAD_REQUEST)
+
+        for collection_id in request.data['ids']:
+            self._check_collection_permissions(collection_id, request.user)
+
+        for collection_id in request.data['ids']:
+            entity.collections.remove(collection_id)
+
+            collection = Collection.objects.get(pk=collection_id)
+            for data in entity.data.all():
+                collection.data.remove(data)
+
+        return Response()
+
+    @detail_route(methods=[u'post'])
+    def add_data(self, request, pk=None):
+        """Add data to Entity and it's collection."""
+        # add data to entity
+        resp = super(EntityViewSet, self).add_data(request, pk)
+
+        # add data to collections in which entity is
+        entity = self.get_object()
+        for collection in entity.collections.all():
+            collection.data.add(*request.data['ids'])
+
+        return resp
+
+    @detail_route(methods=[u'post'])
+    def remove_data(self, request, pk=None):
+        """Remove Data from Entity and delete it if it is empty."""
+        resp = super(EntityViewSet, self).remove_data(request, pk)
+
+        entity = self.get_object()
+        if entity.data.count() == 0:
+            entity.delete()
```

Comment:
Is this needed? There is a signal handler in `signals.py`, which does this automatically?